apps/aecore/src/aec_chain_server.erl

```diff
@@ -0,0 +1,201 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2017, Aeternity Anstalt
+%%% @doc Service holding blocks (or headers) from the block chain.
+%%%
+%%% The longest chain is determined according to the amount of work
+%%% done on the chain, i.e. according to the total difficulty of the
+%%% highest header in the chain (the so-called "top" header).  The
+%%% total difficulty of a header is the sum of the difficulty of the
+%%% header and the total difficulty of the previous header.
+%%%
+%%% The difficulty of a header is a linear representation of the
+%%% expected average amount of work required for mining the header,
+%%% and is derived from the target threshold in the header.
+%%%
+%%% The longest chain is also called the main chain.
+%%% The server will keep the state (account balances) of the main chain.
+%%% These states will be kept (checkpointed) at certain points in the
+%%% chain in order to quickly rebuild the state when merging and
+%%% switching to an alternate chain.
+%%%
+%%% The server also track blocks in alternative chains.
+%%% That is blocks or headers not on the main chain.
+%%%
+%%% If an alternative chain becomes rooted in a different genesis
+%%% block that alternative chain is thrown away.
+%%%
+%%% If an alternative chain has more work than the current chain
+%%% and is rooted in a common ancestor the alternative chain will
+%%% become the main chain.
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+-module(aec_chain_server).
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/1,
+         stop/0]).
+%% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2, code_change/3]).
+
+-include("common.hrl").
+-include("blocks.hrl").
+
+-define(SERVER, ?MODULE).
+
+%%%===================================================================
+%%% API -  the main API for the server is in aec_chain.
+%%%===================================================================
+
+start_link(GenesisBlock) ->
+    Args = [GenesisBlock],
+    gen_server:start_link({local, ?SERVER}, ?MODULE, Args, []).
+
+stop() ->
+    gen_server:stop(?SERVER).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+ 
+init(_Args = [GenesisBlock]) ->
+    process_flag(trap_exit, true),
+    GenesisHeader = aec_blocks:to_header(GenesisBlock),
+    {ok, S2} = insert_header(GenesisHeader, new_state()),
+    {ok, S3} = insert_block(GenesisBlock, S2),
+    {ok, S3}.
+
+
+%% State preserving functions
+handle_call(top, _From, State) ->
+    {reply, top(State), State};
```

Comment:
Recognizing that this implementation is a 'shadow' version of `aec_chain`, is the intention to rename this module as `aec_chain`? It's not common practice to have API functions in one module do direct gen_server calls to a server implemented in another. If API functions are added to this module later, it seems as if the intuitive names for these functions are already reserved for internal functions (though perhaps with different arity). I would suggest using an 'internal naming' scheme for these – either `*_int()` or just `*_()` (which is a convention e.g. Tony Rogvall likes to use.)

Reply:
I think it is a better and cleaner practice to separate the outside API and the server implementation. We can discuss offline, and then come back here for a conclusion.